hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/utils/HiveAvroSerializer.java

```diff
@@ -493,12 +499,16 @@ private static void copyOldValueOrSetDefault(GenericRecord oldRecord, GenericRec
       Object newFieldValue;
       if (fieldValue instanceof GenericRecord) {
         GenericRecord record = (GenericRecord) fieldValue;
-        newFieldValue = rewriteRecordIgnoreResultCheck(record, AvroSchemaUtils.resolveUnionSchema(field.schema(), record.getSchema().getFullName()));
+        HoodieSchema nonNullFieldSchema = field.schema().getNonNullType();
+        if (!Objects.equals(nonNullFieldSchema.getFullName(), record.getSchema().getFullName())) {
```

Comment:
Can you add some details on the logic change here?

Reply:
I'm flattening the logic here for `AvroSchemaUtils#resolveUnionSchema`.

The first part of `AvroSchemaUtils#resolveUnionSchema` is pretty much the same as `HoodieSchema#getNonNullType()` where there's a check on whether the schema is of **UNION** type.

If so, we attempt to extract the type that is not null from the union pair.

The difference here is that `AvroSchemaUtils#resolveUnionSchema` seems to be able to account for UNIONS that have a size larger than 2, which `HoodieSchema#getNonNullType()` does not. 

And so, I felt that this is an unreachable case, which is why i added the sanity check back there as a safeguard.

Reply:
Moving this utility to HoodieSchemaUtils and adding tests.